### PR TITLE
[RISC-V] Fix struct types passed by floating-point calling convention

### PR DESCRIPTION
**Pull request by tomeksowi**
https://github.com/dotnet/runtime/pull/103537#issuecomment-2171755041

Part of #84834, cc [@dotnet/samsung](https://github.com/orgs/dotnet/teams/samsung)

### CODE REVIEW

**Comment by jakobbotsch** [comment]
Thanks for looking at this. Feel free to submit it upstream since it looks like it doesn't interact with any of the things I was doing in this branch.

**Comment by jakobbotsch** [comment]
> I still need to resolve one more assert failure, seemingly unrelated to the stack arg classification (during crossgen2 of System.Numerics.Matrix3x2:setItem(int, int, float):this):
> https://github.com/dotnet/runtime/blob/eb455ec34c6709e487c19e52c29ec712a6fa4d7f/src/coreclr/jit/lsrabuild.cpp#L606-L611

On a related note I think this assert can/should be tightened back to the same as all other targets; it should be possible after the #97368 (related comment: https://github.com/dotnet/runtime/pull/97368#discussion_r1464799403).

**Comment by tomeksowi** [comment]
Should work now, albeit platform-specific fix-ups are not the nicest thing. I phrased it utilizing new ABI passing info to make it easier when the old ones will eventually go.

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -2126,6 +2126,33 @@ void CallArgs::AddFinalArgsAndDetermineABIInfo(Compiler* comp, GenTreeCall* call
...
+                    assert(arg.NewAbiInfo.NumSegments == 1);
+                    structBaseType = arg.NewAbiInfo.Segments[0].GetRegisterType();
+                }
+
+                for (unsigned i = 0; i < arg.NewAbiInfo.NumSegments; ++i)
+                {
+                    arg.AbiInfo.StructFloatFieldType[i] = arg.NewAbiInfo.Segments[i].GetRegisterType();
+                }
+            }
+#endif // defined(TARGET_RISCV64) || defined(TARGET_LOONGARCH64)
```
Comment (tomeksowi) [inline src/coreclr/jit/morph.cpp:2155]: cc @LuckyXu-HF @shushanhf

**Comment by jakobbotsch** [comment]
> Should work now, albeit platform-specific fix-ups are not the nicest thing. I phrased it utilizing new ABI passing info to make it easier when the old ones will eventually go.

Looks good to me. I think this platform-specific fixup will disappear once the multireg morphing code is adapted to use the new ABI representation directly.

File: src/coreclr/jit/morph.cpp (hunk 1 above)
Comment (LuckyXu-HF) [reply src/coreclr/jit/morph.cpp:2155]: LGTM, thanks. I think we need some test further on LA64 after the relate refactor of these.

**Pull request merged**